### ISSUE TITLE
improve and refactor solve_math to not recreate an iterator for every operation inside an equation

### ISSUE THREAD

**Issue by atomicol** [eskevv/cal-rusty#2]
This happens because we are using vectors to store the results of every operation within a loop. Removing elements forces us to either recreate the iterator or to reverse the loop. Reversing won't work well here because it would require switching the lhs and rhs operands specifically for the current order of operations. This can get messy when we factor in the pow operator and just doesn't look nice.

Suggestions:
1. reverse the collection and then reverse the loop
2. use an array instead or just don't remove elements (leave the old results in the collection)
   -> this might require an extra step of creating nested collections of results

**Comment by atomicol** [comment]
Suggestion 2 ended up recreating a collection of results for every level of presedence. These results had to be linked to an index to which it took up space from the original collection of operands. In the end I had to iterate through these indexes to check whether I should operate on the next index of the original collection or a previous calculation from prior presendence; It throws off the whole idea of not wanting to iterate for as many operators and doesn't result in any improvement. 
This means I now have to recreate the original array ( why we just use vectors ) or add some sort of counter to how many indexes we should skip over. This will require a counter for every level of presedence but can be faulty and means extra tests. Or even better use a dictionary of indixes where we can use O(1) to check if that index already holds a result.

**Comment by atomicol** [comment]
I was able to figure out how to go about doing the first one since it seems the most clean, however it was a little more torublesome than I had expected. I did have to reverse the collection and then I needed to iterate from the end, but of course I also needed to account for the operators that were being removed and keep a counter to an index offset. 

* The counter would need to reset every time we increase the operator presedence and increased for every operation.
* The evaluated value of every operation is now stored in the opposite order; That being at the index instead of index + 1
    -> also the reason we needed an offset here e.g. if the first operation was done at index 0 the result would have to be stored now at index 1 (as opposed to 0 if we didnt reverse) because we should remove and store always at the root of evaluation(left-> right) even for exponents -therefore we need indicate it is the first value by applying an offset.
* Finally this ensures we can create a single iterator for any given presedence level while destroying elements from both the operator and operand collection and still maintain a correct order of operations
